tools/tdbg/dlq_v2_service.go

```diff
@@ -0,0 +1,177 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tdbg
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/urfave/cli/v2"
+	"go.temporal.io/server/api/adminservice/v1"
+	"go.temporal.io/server/api/enums/v1"
+	"go.temporal.io/server/common/collection"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+type DLQV2Service struct {
+	category      tasks.Category
+	sourceCluster string
+	targetCluster string
+	clientFactory ClientFactory
+}
+
+func NewDLQV2Service(
+	category tasks.Category,
+	sourceCluster string,
+	targetCluster string,
+	clientFactory ClientFactory,
+) *DLQV2Service {
+	return &DLQV2Service{
+		category:      category,
+		sourceCluster: sourceCluster,
+		targetCluster: targetCluster,
+		clientFactory: clientFactory,
+	}
+}
+
+func (ac *DLQV2Service) ReadMessages(c *cli.Context) error {
+	ctx, cancel := newContext(c)
+	defer cancel()
+
+	remainingMessageCount := math.MaxInt32
+	if c.IsSet(FlagMaxMessageCount) {
+		remainingMessageCount = c.Int(FlagMaxMessageCount)
+		if remainingMessageCount <= 0 {
+			return fmt.Errorf("--%s must be positive but was %d", FlagMaxMessageCount, remainingMessageCount)
+		}
+	}
+	var maxMessageID int64 = math.MaxInt64
+	if c.IsSet(FlagLastMessageID) {
+		maxMessageID = c.Int64(FlagLastMessageID)
+		if maxMessageID < persistence.FirstQueueMessageID {
+			return fmt.Errorf(
+				"--%s must be at least %d but was %d",
+				FlagLastMessageID,
+				persistence.FirstQueueMessageID,
+				maxMessageID,
+			)
+		}
+	}
+
+	outputFile, err := getOutputFile(c.String(FlagOutputFilename))
+	if err != nil {
+		return err
+	}
+	adminClient := ac.clientFactory.AdminClient(c)
+
+	pageSize := c.Int(FlagPageSize)
+	iterator := collection.NewPagingIterator[*adminservice.HistoryDLQTask](
+		func(paginationToken []byte) ([]*adminservice.HistoryDLQTask, []byte, error) {
+			request := &adminservice.GetDLQTasksRequest{
+				DlqKey: &adminservice.HistoryDLQKey{
+					Category:      enums.TaskCategory(ac.category.ID()),
+					SourceCluster: ac.sourceCluster,
+					TargetCluster: ac.targetCluster,
+				},
+				PageSize:      int32(pageSize),
+				NextPageToken: paginationToken,
+			}
+			res, err := adminClient.GetDLQTasks(ctx, request)
+			if err != nil {
+				return nil, nil, fmt.Errorf("call to GetDLQTasks from ReadMessages failed: %w", err)
+			}
+			return res.DlqTasks, res.NextPageToken, nil
+		},
+	)
+
+	encoder := jsonpb.Marshaler{
+		Indent:       "  ",
+		EmitDefaults: true,
+	}
+	for iterator.HasNext() && remainingMessageCount > 0 {
+		dlqTask, err := iterator.Next()
+		if err != nil {
+			return fmt.Errorf("DLQ task iterator returned error: %w", err)
+		}
+		if dlqTask.Metadata.MessageId > maxMessageID {
+			break
+		}
+		remainingMessageCount--
+		taskString, err := encoder.MarshalToString(dlqTask)
```

Comment:
this will print out raw task blob bytes?

Reply:
For now, yes. I added a TODO because I still haven't figured out how this will work with dynamic task categories.